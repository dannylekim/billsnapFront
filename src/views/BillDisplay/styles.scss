@import '../../variables';

div.bill__wrapper{
    height:500px;
    min-width: 1313px;
}

div.bill__section{
    display: flex;
}

div.specific__bill__section{
    width: 61.091%;

    span#more__details{
        text-decoration: underline;
        color:$primary-text-color;
        float: right;
        margin-bottom: 15px;
        margin-right: 32px;
        right:0;
    }

    h5 span#amount__owed{
        color:$secondary-price-color;
    }
}

div.bill__list__section{
    margin-left: 2.74117026885%;
    width: 38.909%;
  
    #search__bill{
        border: 1px solid $light-gray-background;
        box-sizing: border-box;
        border-radius: 50px;
        background: $light-gray-background;
        display: flex;
        align-content: center;
        margin-bottom: 25px;
    }

    span.search__icon{
        width: 10%;
        border-top-left-radius: inherit;
        border-bottom-left-radius: inherit;
        border: inherit;
        background: inherit;
        margin-top: auto;
        margin-bottom: auto;
    }
    
    span.search__filter{
        width: 30%;
        margin-top: auto;
        margin-bottom: auto;
        font-weight: 500;
        font-size: 20px;
        border-top-right-radius: inherit;
        border-bottom-right-radius: inherit;
        span.simple__sort {
            margin-left: 10px;
            
        }
    }

    input.form-control{
        width: 60%;
        border: inherit;
        background: inherit;
        box-shadow: none;
    }

    ul.nav.nav-tabs{
        justify-content: start;
        border-bottom: 1px solid $border-color;
        margin-bottom: 15px;
    }

    button#add__bill__button{
        margin-left: calc(100% - 140px);
        margin-top: 10px;
        margin-bottom: 10px;
        border-radius: 17px;
        font-weight: bold;
        font-size: 20px;
        border: 1px solid $secondary-color;
        min-height:40px;
        width: 140px;
    }
    a.nav-link{
        height:50px;
        width: 140px;
        border-radius: 10px 10px 0px 1px;
        margin-left: 5px;
        font-size: 16px;
        border-bottom: 1px solid $border-color;
    }

    a.nav-link.active{
      border: 1px solid $border-color;
      border-bottom: none;
    }

    button#add__bill__button,a.nav-link{
        justify-content : end;
        background-color: $secondary-color;
        color: $card-color;
        box-sizing: border-box;
        box-shadow: 4px 0px 4px $card-shadow;
        font-family: $primary-font;
        font-style: normal;
        font-weight: 300;
        padding:10px 0;
    }

    div.bill__card.card{
        background-color: $card-color;
        border-radius: 0;
        box-shadow: none;
        height:160px;
    
        hr.card__seperator{
            margin-bottom: 0;
            margin-top: 0;
            background-color:rgba(0, 0, 0, 0.63);
            width: 80.67%;
            margin-left: auto;
            margin-right: auto;
        }
    }

    div.quick__sorting__container{
        position: absolute;
        top:75px;
        height: 120px;
        ul.sorting__content{
            width: 103px;
            transform: translateX(470px);
            border-radius: 6px;
            height: inherit;
            background-color: $light-gray-background;
            list-style : none;
            padding-top:10px;
            padding-left:0;

            li.sorting__titles{
                font-family: $primary-font;
                font-weight: 500;
                font-size: 20px;
                line-height: 26px;
                color: $secondary-text-color;
            }

            li.sorting__titles:hover{
                background-color: #bebebe;
            }
        }
    }

    div.bill__container{
        display: flex;
        flex-wrap: nowrap;
        flex-direction: column;
        margin-bottom: 15px;
        box-shadow: -1px -1px 2px rgba(90, 97, 105, 0.25);
    }
    
    div.bill__summary{
        text-align: left;
        margin-left: 130px;
    }  
}

