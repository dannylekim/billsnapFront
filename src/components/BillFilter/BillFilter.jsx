import React, { useRef } from "react";
import { Badge, FormCheckbox, FormInput } from "shards-react";

import { useOutsideAlerter } from "../../helpers/ClickEvent";
import "./styles.scss";

/**
 * @description the filter box when three horizontal lines clicked
 */
const BillFilter = ({
  activeTab,
  applyFiltering,
  currentSorting,
  billStatusFilter,
  closeHandler,
  dateCheckboxHandler,
  dateFilters,
  filterToggles,
  filterToggleChange,
  handleDateSelection,
  handleStatusChange,
}) => {
  const wrapperRef = useRef(null);
  useOutsideAlerter(wrapperRef, closeHandler);

  const filterBadges = [
    {
      id: "descending",
      title: "Newest",
      onClick: () => applyFiltering("Newest"),
    },
    {
      id: "ascending",
      title: "Oldest",
      onClick: () => applyFiltering("Oldest"),
    },
    {
      id: "by__status",
      title: "Status",
      onClick: () =>
        filterToggleChange({
          statusOpened: true,
          categoryOpened: false,
          dateOpened: false,
        }),
    },
    {
      id: "alphabetical__increasing",
      title: "A to Z",
      onClick: () => applyFiltering("A to Z"),
    },
    {
      id: "alphabetical__decreasing",
      title: "Z to A",
      onClick: () => applyFiltering("Z to A"),
    },
    {
      id: "by__category",
      title: "Category",
      onClick: () =>
        filterToggleChange({
          categoryOpened: true,
          statusOpened: false,
          dateOpened: false,
        }),
    },
    {
      id: "by__date",
      title: "Date",
      onClick: () =>
        filterToggleChange({
          dateOpened: true,
          categoryOpened: false,
          statusOpened: false,
        }),
    },
  ];

  const extraFilterSelected =
    filterToggles.dateOpened === true ||
    filterToggles.statusOpened === true ||
    filterToggles.categoryOpened === true;
  
  return (
    <div className='bill__filter__container' ref={wrapperRef}>
      <div className='filter__header'>
  <span className='filter__header__title'>Filters {activeTab}</span>
      </div>
      <div className='filter__selections__container'>
        {filterBadges.map((badgeObject, key) => (
          <Badge
            key={key}
            className={badgeObject.title === currentSorting && 'active'}
            outline
            pill
            id={badgeObject.id}
            onClick={badgeObject.onClick}>
            {badgeObject.title}
          </Badge>
        ))}
      </div>
      {extraFilterSelected && (
        <div className='advanced__filter__selection'>
          {filterToggles.dateOpened === true && (
            <div className='date__filter'>
              <FormCheckbox
                checked={dateFilters.startDate.selected}
                onChange={() => dateCheckboxHandler("startDate")}>
                Start date
                {dateFilters.startDate.value !== ""
                  ? `: ${dateFilters.startDate.value}`
                  : ""}
              </FormCheckbox>
              <FormCheckbox
                checked={dateFilters.endDate.selected}
                onChange={() => dateCheckboxHandler("endDate")}>
                End date
                {dateFilters.endDate.value !== ""
                  ? `: ${dateFilters.endDate.value}`
                  : ""}
              </FormCheckbox>
              <FormInput
                type='date'
                onChange={(event) => handleDateSelection(event)}
                disabled={
                  dateFilters.endDate.selected === false &&
                  dateFilters.startDate.selected === false
                }
              />
            </div>
          )}
          {filterToggles.statusOpened === true && (
            <div className='status__filter'>
              <FormCheckbox
                checked={billStatusFilter.resolved}
                onChange={() =>
                  handleStatusChange('resolved')
                }>
                PAID
              </FormCheckbox>
              <FormCheckbox
                checked={billStatusFilter.open}
                onChange={() =>
                  handleStatusChange('open')
                }>
                UNPAID
              </FormCheckbox>
              <FormCheckbox
                checked={billStatusFilter.in_progess}
                onChange={() =>
                  handleStatusChange('in_progess')
                }>
                ONGOING
              </FormCheckbox>
            </div>
          )}
          {filterToggles.categoryOpened === true && (
            <div className='category__filter'>
              <FormInput type='text' placeholder='Restaurant' />
            </div>
          )}
        </div>
      )}
    </div>
  );
};

export default BillFilter;
