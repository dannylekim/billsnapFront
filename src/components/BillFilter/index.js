import BillFilter from "./BillFilter.jsx";

export default BillFilter;
