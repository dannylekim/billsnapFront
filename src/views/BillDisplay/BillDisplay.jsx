import React, {Component} from "react";
import "./styles.scss";
// import Loader from "../../components/Loader";
import SmallBillCard from "../../components/BillCard/SmallBillCard";
import LargeBillCard from "../../components/BillCard/LargeBillCard";
import BillFilter from "../../components/BillFilter";
import {Button, Nav, NavItem, NavLink} from "shards-react";
import {FaBars, FaBus, FaCar, FaQuestion, FaSearch, FaShoppingBag, FaShoppingCart, FaUtensils,} from "react-icons/fa";
import navItems from "../../constants/BillDisplayNav.json";

class BillDisplay extends Component {
  constructor(props) {
    super(props);
    this.state = {
      seachedQuery: "",
      selectedBill: { bill: null, id: 0 },
      currentActiveTab: "allBills",
      sorting: { opened: false, type: "Newest" },
      filter: {
        opened: false,
        type: "",
        statusOpened: false,
        categoryOpened: false,
        dateOpened: false,
      },
      dateFilters: {
        startDate: { selected: false, value: "" },
        endDate: { selected: false, value: "" },
      },
      billStatusFilter: { resolved: false, open: false, in_progess: false },
    };
  }

  componentDidMount = async () => {
    await this.props.fetchBills();
  };

  static billIcons = (category) => {
    const color = "rgba(0, 0, 0, 0.96)";
    switch (category) {
      case "food":
        return <FaUtensils color={color} size={24} />;
      case "transport":
        return <FaCar color={color} size={24} />;
      case "public-transport":
        return <FaBus color={color} size={24} />;
      case "grocery":
        return <FaShoppingCart color={color} size={24} />;
      case "shopping":
        return <FaShoppingBag color={color} size={24} />;
      default:
        return <FaQuestion color={color} size={24} />;
    }
  };

  static filterDateTime = (timestamp) => {
    let dateTime;
    const date = timestamp.split(" ")[0];
    const time = timestamp.split(" ")[1];
    const dateArray = date.split("-");
    const year = parseInt(dateArray[2]);
    const month = parseInt(dateArray[1]) - 1;
    const day = parseInt(dateArray[0]);

    const dateCreated = new Date(year, month, day).setHours(0, 0, 0, 0);
    const todayDate = new Date().setHours(0, 0, 0, 0);
    if (dateCreated < todayDate) {
      dateTime =
        year < new Date(todayDate).getFullYear()
          ? `${day < 10 ? "0" + day : day}/${
              month + 1 < 10 ? "0" + (month + 1) : month + 1
            }/${year}`
          : `${day < 10 ? "0" + day : day}/${
              month + 1 < 10 ? "0" + (month + 1) : month + 1
            }`;
    } else {
      const hour = parseInt(time.split(":")[0]);
      const minute = parseInt(time.split(":")[1]);
      const AMPM = hour >= 12 ? "PM" : "AM";
      dateTime = `${
        hour > 12 || hour === 0 ? Math.abs(hour - 12) : hour
      }:${minute} ${AMPM}`;
    }
    return dateTime;
  };

  updateBills = (type) => {
    this.setState({
      sorting: { type, opened: false },
      filter: {
        ...this.state.filter,
        opened: false,
        categoryOpened: false,
        statusOpened: false,
        dateOpened: false,
      },
      dateFilters: {
        ...this.state.dateFilters,
        startDate: { ...this.state.dateFilters.startDate, selected: false },
        endDate: { ...this.state.dateFilters.endDate, selected: false },
      },
    });
    if (type !== "A to Z" && type !== "Z to A") {
      const startDate = `start=${this.state.dateFilters.startDate.value}`;
      const endDate = `end=${this.state.dateFilters.endDate.value}`;

      const filterQueryParam = {
        Newest: `?${startDate}&${endDate}&sort_by=CREATED&order_by=DESC`,
        Oldest: `?${startDate}&${endDate}&sort_by=CREATED&order_by=ASC`,
      };
      this.props.fetchBills(filterQueryParam[type]);
    } else this.props.orderAlphabetical(type, this.props.bills);
  };

  handleDateSelection = (event) => {
    event.preventDefault();
    const date = event.target.value;

    if (date !== "") {
      if (
        this.state.dateFilters.startDate.selected === true &&
        this.state.dateFilters.endDate.selected === false
      ) {
        this.setState({
          dateFilters: {
            ...this.state.dateFilters,
            startDate: { ...this.state.dateFilters.startDate, value: date },
          },
        });
      }

      if (
        this.state.dateFilters.endDate.selected === true &&
        this.state.dateFilters.startDate.selected === false
      ) {
        this.setState({
          dateFilters: {
            ...this.state.dateFilters,
            endDate: { ...this.state.dateFilters.endDate, value: date },
          },
        });
      }
    } else {
      return null;
    }
  };

  render = () => {
    const { bills, isBillLoading, count } = this.props;

    /**
     * @description the search bar of the bill
     */
    const searchBill = (
      <div id="search__bill">
        <span className="search__icon">
          <FaSearch />
        </span>
        <input
          type="text"
          className="form-control border-0"
          onChange={(event) =>
            this.setState({ seachedQuery: event.target.value })
          }
          placeholder="Search bill"
        />
        <span className="search__filter">
          <span
            className="filter__bill"
            onClick={() =>
              this.setState({
                filter: {
                  ...this.state.filter,
                  opened: !this.state.filter.opened,
                },
                sorting: {
                  ...this.state.sorting,
                  opened: false,
                },
              })
            }
          >
            <FaBars size={24} />
          </span>
          <span
            className="simple__sort"
            onClick={() =>
              this.setState({
                filter: {
                  ...this.state.filter,
                  opened: false,
                },
                sorting: {
                  ...this.state.sorting,
                  opened: !this.state.sorting.opened,
                },
              })
            }
          >
            {this.state.sorting.type}{" "}
          </span>
        </span>
      </div>
    );

    /**
     * @description the sorting box when furthest right is clicked
     */
    const simpleSort = (
      <div className="quick__sorting__container">
        <ul className="sorting__content">
          {["Newest", "A to Z", "Z to A", "Oldest"]
            .filter((title) => title !== this.state.sorting.type)
            .map((title, key) => (
              <li
                className="sorting__titles"
                onClick={() => this.updateBills(title)}
                key={key}
              >
                {title}
              </li>
            ))}
        </ul>
      </div>
    );

    /**
     * @description the add new bill button (add new bill)
     */
    const AddBillButton = (
      <Button id="add__bill__button"> {"+ Add bill"} </Button>
    );

    const resetFilters = () => {
      this.props.fetchBills();
      this.setState({
        sorting: { opened: false, type: "Newest" },
        dateFilters: {
          startDate: { selected: false, value: "" },
          endDate: { selected: false, value: "" },
        },
        billStatusFilter: { resolved: false, open: false, in_progess: false },
      });
    };

    const NavigationTabs = (tabItems) => (
      <Nav tabs>
        {tabItems.map((item, key) => (
          <NavItem key={key} onClick={() => key === 0 && resetFilters()}>
            <NavLink
              active={this.state.currentActiveTab === item.name}
              onClick={() => this.setState({ currentActiveTab: item.name })}
            >
              {item.title}
            </NavLink>
          </NavItem>
        ))}
      </Nav>
    );
    /**
     * @description returns the list of bills as cards.
     * @param {Array} billsVar the variable bills,
     */
    const BillsList = (billsVar) => {
      billsVar =
        this.state.seachedQuery.trim() !== ""
          ? billsVar.filter((bill) =>
              bill.name.includes(this.state.seachedQuery)
            )
          : billsVar;

      return (
        <div className="bill__container">
          {billsVar.length > 0 ? (
            billsVar.map((bill, key) => (
              <div
                className="bill__card card"
                key={key}
                onClick={() =>
                  this.setState({ selectedBill: { bill, id: key + 1 } })
                }
              >
                <SmallBillCard
                  activeBill={this.state.selectedBill.bill}
                  bill={bill}
                  filterDateTime={this.constructor.filterDateTime}
                  billIcons={this.constructor.billIcons}
                />
                {key !== billsVar.length - 1 && (
                  <hr className="card__seperator" />
                )}
              </div>
            ))
          ) : (
            <p> {`No bills found titled: ${this.state.seachedQuery}`}</p>
          )}
        </div>
      );
    };

    const BillsSummary = (billsVar) => {
      return this.state.selectedBill.bill === null ? (
        <div className="bill__summary">
          <h5>
            {" "}
            Total Amount Owed :{" "}
            <span id="amount__owed">
              {" "}
              {billsVar
                .reduce((a, b) => a + parseFloat(b.balance), 0)
                .toFixed(2)}{" "}
              $
            </span>{" "}
          </h5>
          <h5> Total of bills : {count} </h5>
        </div>
      ) : (
        <div className="bill__summary">
          <h5>
            {`Split by : ${this.state.selectedBill.bill.responsible.firstName} ${this.state.selectedBill.bill.responsible.lastName}`}
          </h5>
          <h5> {`Status : ${this.state.selectedBill.bill.status}`}</h5>
          <h5>
            Amount Owed :{" "}
            <span id="amount__owed">
              {this.state.selectedBill.bill.balance} $
            </span>{" "}
          </h5>
        </div>
      );
    };

    const { dateFilters, filter, billStatusFilter } = this.state;
    return (
      <div className="bill__wrapper">
        {isBillLoading ? (
          // <Loader />
          "nice"
        ) : bills.length > 0 ? (
          <div className="bill__section">
            <div className="bill__list__section">
              {searchBill}
              {AddBillButton}
              {NavigationTabs(navItems)}
              {BillsList(bills)}
              <BillFilter
                dateFilters={dateFilters}
                filter={filter}
                billStatusFilter={billStatusFilter}
                handleDateSelection={this.handleDateSelection}
                updateBills={this.updateBills}
                setState={this.setState.bind(this)}
              />
              {this.state.sorting.opened === true && simpleSort}
            </div>
            <div className="specific__bill__section">
              {BillsSummary(bills)}
              <span id="more__details"> More details</span>
              <LargeBillCard selectedBill={this.state.selectedBill} />
            </div>
          </div>
        ) : (
          <p> No Bills Found</p>
        )}
      </div>
    );
  };
}

export default BillDisplay;
