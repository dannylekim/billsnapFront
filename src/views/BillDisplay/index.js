import BillDisplay from "./BillDisplay.jsx";

import {connect} from "react-redux";
import {fetchMyBills, orderAlphabetical,} from "../../redux/actions/billActions";

const mapStateToProps = (state) => ({
  count: state.bills.count,
  bills: state.bills.bills,
  isBillLoading: state.application.isBillLoading,
});

const mapDispatchToProps = (dispatch) => ({
  fetchBills: (queryParam = "") => dispatch(fetchMyBills(queryParam)),
  orderAlphabetical: (alphabeticalType, bills) =>
    dispatch(orderAlphabetical(alphabeticalType, bills)),
});

export default connect(mapStateToProps, mapDispatchToProps)(BillDisplay);
