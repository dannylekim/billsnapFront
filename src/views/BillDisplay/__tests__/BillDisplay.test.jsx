import React from "react";
import BillDisplay from "../BillDisplay";
import BillFilter from "../../../components/BillFilter";
import {shallow} from "enzyme";
import {FaBus, FaCar, FaQuestion, FaShoppingBag, FaShoppingCart, FaUtensils} from 'react-icons/fa';

jest.mock("../../../utils/requests/BillRequests", () => {
  return {
    getBills: jest.fn().mockResolvedValue(["bill1", "bill2"]),
  };
});

describe("BillDisplay", () => {
      let wrapper, instance;
      let mockFetch = jest.fn(),mockSortAlphabetical = jest.fn();

      const responsible = ({
        id: 1,
        email: "someEmail@email.com",
        firstName: "Bob",
        middleName: null,
        lastName: "Smith",
        gender: null,
        phoneNumber: null,
        birthDate: null,
        location: null
      });

      const mockBills = [
        {
          id: 1,
          name: "BILL MOCK 1",
          status: "OPEN",
          category: "string",
          created:  "05-03-2018 15:25:10 -0400",
          balance: 12.0,
          responsible
        },
        {
          id: 2,
          name: "BILL MOCK 2",
          status: "OPEN",
          category: "string",
          created:  "06-04-2018 15:25:10 -0400",
          balance: 15.0,
          responsible
        },
      ];

      beforeEach(() => {
        wrapper = shallow(
          <BillDisplay bills={mockBills} fetchBills={mockFetch} isBillLoading={false} orderAlphabetical={mockSortAlphabetical} />
        );

        instance = wrapper.instance();
      });
      afterEach(() => {
        mockFetch.mockRestore();
        mockSortAlphabetical.mockRestore();
      });

      describe("render", () => {
          describe("snapshots ðŸ“¸", () => {
            it("BillDisplay should match snap shot when loading", () => {
              matches(
                <BillDisplay bills={[]} fetchBills={mockFetch} isBillLoading={true} orderAlphabetical={mockSortAlphabetical} />
              );
            });

            it("BillDisplay should match snap shot when done loading + bills", () => {
              matches(
                <BillDisplay bills={mockBills} fetchBills={mockFetch} isBillLoading={false} orderAlphabetical={mockSortAlphabetical} />
              );
            });

            it("BillDisplay should match snap shot when done loading + no bills", () => {
              matches(
                <BillDisplay bills={[]} fetchBills={mockFetch} isBillLoading={false} orderAlphabetical={mockSortAlphabetical} />
              );
            });
          });
      });

      describe("components", () => {
       
        it("should change seachedQuery state on form search.", () => {
          const event = {
            target: { value: 'hahahhahahahha' }
          };
         
          wrapper.find('input.form-control.border-0').simulate('change', event);
          expect(wrapper.state().seachedQuery).toBe(event.target.value);
        })

        it("should change selectedBill on click.", () => {
          wrapper.find('div.bill__card.card').at(0).simulate('click');
          expect(wrapper.state().selectedBill.bill).toBe(mockBills[0]);
          wrapper.find('div.bill__card.card').at(1).simulate('click');
          expect(wrapper.state().selectedBill.bill).toBe(mockBills[1]);
        })

        it("should change activeTab on click.", () => {
          expect(wrapper.state().currentActiveTab).toBe("allBills");
          wrapper.find('NavLink').at(1).simulate('click');
          expect(wrapper.state().currentActiveTab).toBe("owedToYou");
        });

        it("should call mock function on all bills", () => {
          wrapper.find('NavItem').at(0).simulate('click');
          expect(mockFetch).toHaveBeenCalled();
        });

        describe("simple sort onClicks", () => {
          beforeEach(() => {
            wrapper.find('span.simple__sort').simulate('click');
            wrapper.update();
          });

          it("should trigger a Mock function when A to Z is clicked.", () => {
            wrapper.find('li.sorting__titles').at(0).simulate('click');
            expect(mockSortAlphabetical).toHaveBeenCalled()
            expect(wrapper.state().sorting.type).toBe("A to Z");
          })

          it("should trigger a Mock function when Z to A is clicked.", () => {
            wrapper.find('li.sorting__titles').at(1).simulate('click');
            expect(mockSortAlphabetical).toHaveBeenCalled()
            expect(wrapper.state().sorting.type).toBe("Z to A");
          })

          it("should trigger a Mock function when Newest is selected after oldest is selected.", () => {
            wrapper.find('li.sorting__titles').at(2).simulate('click');
            expect(mockFetch).toHaveBeenCalled()
            wrapper.find('span.simple__sort').simulate('click');
            wrapper.update();
            wrapper.find('li.sorting__titles').at(0).simulate('click');
            expect(mockFetch).toHaveBeenCalled()
          });
        });

        describe("filter badge onClicks", () => {
          let billFilterWrapper, billWrapper, mockHandleDateSelection;
          beforeEach(() => {
            expect(wrapper.state().filter.opened).toBeFalsy();
            wrapper.find('span.filter__bill').simulate('click');
            wrapper.update();

            mockHandleDateSelection = jest.fn();
            billFilterWrapper = () => shallow(<BillFilter dateFilters={wrapper.state().dateFilters}
                                                          filter={wrapper.state().filter} 
                                                          billStatusFilter={wrapper.state().billStatusFilter}  
                                                          handleDateSelection={mockHandleDateSelection}
                                                          updateBills={(type) => instance.updateBills(type)}
                                                          setState={(newState) => instance.setState(newState)} />);
            billWrapper = billFilterWrapper();
          });

          it("should close filter div", () => {
            expect(wrapper.state().filter.opened).toBeTruthy();
            billWrapper.find('div.filter__header').simulate('click');
            wrapper.update();
            expect(wrapper.state().filter.opened).toBeFalsy();
          });
          
          it("should trigger a Mock function when Newest is selected.", () => {
            billWrapper.find('Badge').at(0).simulate('click');//calls updateBills
            expect(mockFetch).toHaveBeenCalled();
            expect(wrapper.state().sorting.type).toBe("Newest");
          });

          it("should trigger a Mock function when Oldest is clicked.", () => {
            billWrapper.find('Badge').at(1).simulate('click');
            expect(mockFetch).toHaveBeenCalled()
            expect(wrapper.state().sorting.type).toBe("Oldest");
          });
          
          it("should change status checkbox states on status filter select.", () => {
            expect(wrapper.state().filter.statusOpened).toBeFalsy();
            billWrapper.find('Badge').at(2).simulate('click');
            expect(wrapper.state().filter.statusOpened).toBeTruthy();
            
            billWrapper = billFilterWrapper();

            billWrapper.find('FormCheckbox').at(2).simulate('change');
            wrapper.update();
            expect(wrapper.state().billStatusFilter.in_progess).toBeTruthy();
            expect(wrapper.state().billStatusFilter.resolved).toBeFalsy();
            expect(wrapper.state().billStatusFilter.open).toBeFalsy();
            
            billWrapper.find('FormCheckbox').at(1).simulate('change');
            expect(wrapper.state().billStatusFilter.in_progess).toBeFalsy();
            expect(wrapper.state().billStatusFilter.resolved).toBeFalsy();
            expect(wrapper.state().billStatusFilter.open).toBeTruthy();
            
            billWrapper.find('FormCheckbox').at(0).simulate('change');
            expect(wrapper.state().billStatusFilter.in_progess).toBeFalsy();
            expect(wrapper.state().billStatusFilter.resolved).toBeTruthy();
            expect(wrapper.state().billStatusFilter.open).toBeFalsy();
          });

          it("should trigger a Mock function when A to Z is clicked.", () => {
            billWrapper.find('Badge').at(3).simulate('click');
            expect(mockSortAlphabetical).toHaveBeenCalled()
            expect(wrapper.state().sorting.type).toBe("A to Z");
          });

          it("should trigger a Mock function when Z to A is clicked.", () => {
            billWrapper.find('Badge').at(4).simulate('click');
            expect(mockSortAlphabetical).toHaveBeenCalled()
            expect(wrapper.state().sorting.type).toBe("Z to A");
          });

          it("should trigger category open state on category filter select", () => {
            expect(wrapper.state().filter.categoryOpened).toBeFalsy();
            billWrapper.find('Badge').at(5).simulate('click');
            expect(wrapper.state().filter.categoryOpened).toBeTruthy();
          });

          describe("testing the date function", () => { 
           //has to re-shallow bill filter every time wrapper is updated (.update does not work on functional)
            beforeEach(() => {
              expect(wrapper.state().dateFilters.startDate.selected).toBeFalsy();
              expect(wrapper.state().dateFilters.endDate.selected).toBeFalsy();
             
              billWrapper.find('Badge').at(6).simulate('click');
            
              expect(wrapper.state().filter.dateOpened).toBeTruthy();
            });
            
              describe(" Start date and End date should be triggered on handleDateSelection", () => {
                const event = ({
                  preventDefault: jest.fn(),
                  target: {
                      value: '2020-01-01'
                  }
                });

                it("should change start date checkbox states onChange.", () => {
                  billFilterWrapper().find('FormCheckbox').at(0).simulate('change');
                  wrapper.update();
                  
                  billWrapper =  billFilterWrapper()
                  billWrapper.find('FormInput').simulate('change',event);
                  const spy = jest.spyOn(instance, 'handleDateSelection'); // handleDateSelection
                  instance.handleDateSelection(event);
                  wrapper.update();
                  expect(wrapper.state().dateFilters.startDate.value).toBe(event.target.value);

                  expect(spy).toBeCalledTimes(1);
                  expect(spy).toHaveBeenCalledWith(event);
                  spy.mockRestore();
              });

                it("should change enddate checkbox states onChange.", () => {
                  billFilterWrapper().find('FormCheckbox').at(1).simulate('change');
                  wrapper.update();
                  
                  billWrapper =  billFilterWrapper()
          
                  billWrapper.find('FormInput').simulate('change',event);
                  instance.handleDateSelection(event);
                  wrapper.update();
                  expect(wrapper.state().dateFilters.endDate.value).toBe(event.target.value);
              });
            });

            it("should return null", ()=> {
              const event = ({
                preventDefault: jest.fn(),
                target: {
                    value: ""
                }
              })
              billFilterWrapper().find('FormCheckbox').at(1).simulate('change');
              wrapper.update();
              
              billFilterWrapper().find('FormInput').at(0).simulate('change',event);
              expect(instance.handleDateSelection(event)).toBe(null);
            });
          });
        });
      });

      describe("functions", () => {
          it("fetchBill should be called 1 time", () => {
            expect(mockFetch).toBeCalledTimes(1);
          });

          it("filterDateTime should return correct format", () => {
           
            const today = new Date(new Date().setHours(0,0,0,0));
            expect(BillDisplay.filterDateTime("05-03-2017 15:25:10 -0400")).toBe("05/03/2017");
            expect(BillDisplay.filterDateTime("15-12-2017 15:25:10 -0400")).toBe("15/12/2017");
            expect(BillDisplay.filterDateTime("11-10-2017 15:25:10 -0400")).toBe("11/10/2017");
            expect(BillDisplay.filterDateTime(`01-01-${today.getFullYear()} 15:25:10 -0400`)).toBe("01/01");
            expect(BillDisplay.filterDateTime(`15-01-${today.getFullYear()} 15:25:10 -0400`)).toBe("15/01");
            expect(BillDisplay.filterDateTime(`${today.getDate()}-${today.getMonth() + 1}-${today.getFullYear()} 15:25:10 -0400`)).toBe("3:25 PM");
            expect(BillDisplay.filterDateTime(`${today.getDate()}-${today.getMonth() + 1}-${today.getFullYear()} 3:25:10 -0400`)).toBe("3:25 AM");
          });

          it("fetchBill should return correct icon", () => {
            const color = "rgba(0, 0, 0, 0.96)";
            expect(BillDisplay.billIcons("food")).toStrictEqual(<FaUtensils color={color} size={24}/>);
            expect(BillDisplay.billIcons("transport")).toStrictEqual(<FaCar color={color} size={24}/>);
            expect(BillDisplay.billIcons("public-transport")).toStrictEqual(<FaBus color={color} size={24}/>);
            expect(BillDisplay.billIcons("grocery")).toStrictEqual(<FaShoppingCart color={color} size={24}/>);
            expect(BillDisplay.billIcons("shopping")).toStrictEqual(<FaShoppingBag color={color} size={24}/>);
            expect(BillDisplay.billIcons("unknown")).toStrictEqual(<FaQuestion color={color} size={24}/>);
          });
      });
});