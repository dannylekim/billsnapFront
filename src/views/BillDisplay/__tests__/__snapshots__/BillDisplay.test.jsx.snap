// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillDisplay render snapshots ðŸ“¸ BillDisplay should match snap shot when done loading + bills 1`] = `
<div
  className="bill__wrapper"
>
  <section>
    <h1>
       Bills 
    </h1>
    <div
      className="bill__container"
    >
      <div
        className="bill__card card"
      >
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            BILL MOCK 1
          </h5>
          <ul
            className="bill__list list-group card-text"
          >
            <li
              className="bill__list__item list-group-item list-group-item-success"
            >
              Bill Status: 
              OPEN
            </li>
            <li
              className="bill__list__item list-group-item"
            >
              Bill category: 
              string
            </li>
            <li
              className="bill__list__item list-group-item"
            >
              Bill balance: $
              12
            </li>
          </ul>
        </div>
      </div>
      <div
        className="bill__card card"
      >
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            BILL MOCK 2
          </h5>
          <ul
            className="bill__list list-group card-text"
          >
            <li
              className="bill__list__item list-group-item list-group-item-success"
            >
              Bill Status: 
              OPEN
            </li>
            <li
              className="bill__list__item list-group-item"
            >
              Bill category: 
              string
            </li>
            <li
              className="bill__list__item list-group-item"
            >
              Bill balance: $
              15
            </li>
          </ul>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`BillDisplay render snapshots ðŸ“¸ BillDisplay should match snap shot when done loading + no bills 1`] = `
<div
  className="bill__wrapper"
>
  <p>
     No Bills Found
  </p>
</div>
`;

exports[`BillDisplay render snapshots ðŸ“¸ BillDisplay should match snap shot when loading 1`] = `
<div
  className="bill__wrapper"
>
  <div
    className="billsnap-loader loader"
  >
    Loading...
  </div>
</div>
`;
